jupyterhub/values.yaml

```diff
@@ -35,11 +35,11 @@ hub:
     annotations: {}
     ports:
       nodePort:
-    loadBalancerIP:
+    loadBalancerIP: ""
```

Comment:
Trying to understand why some vars are empty string and some are unset. My intuition is that default should be either a value we want (like `uid: 1000`), unset, or empty container (`[]`, `{}`, etc)

Reply:
The principle ive had while writing these evolved to be:

1: if the helm chart provide an explicit default value: set it.
2: else if the helm chart config is a passthrough config for
jupyterhub/kubespawner/authenticator etc: we set it to null to represent the intent for the helm chart to not change the default value of the underlying config, and a falsy value to represent a wish for it to be explicitly set to falsy value.
3: else if, null snd a falsy instance type both imply a unset value, set it to the falsy instance type
4: else set it to null

To me, the most controversial point is the third. The question in my mind boils down to: if both null and a falsy type ("", {}, [], 0) by the helm templates imply that a value wont be set, should we specify the values falsy type of null?

I lean towards specifying the falsy instance type rather than null if the config isn't passthrough.

I think we break this logic on point 2 for array/objects passthrough config, which i think is motivated by avoiding warnings and help helm templste logic.

Reply:
So with loadBalancerIP, I see the following code

```
  {{- with .Values.proxy.service.loadBalancerIP }}
  loadBalancerIP: {{ . }}
  {{- end }}
```

So if I understand correctly, the `loadBalancerIP` key in the Service object won't be present at all in the following circumstances:

1. `.Values.proxy.service.loadBalancerIP` is specified in helm chart values but not set (current master situation)
2. `.Values.proxy.service.loadBalancerIP` is set to `""` (as in the PR)

Is this understanding correct?

Reply:
Yepp! The with keyword using a falsy value is like a do nothing statement, it has become a practice to use it throughout the helm ecosystem over if ... Then ... - probably to avoid duplicating the reference needed in both the conditional statement and within

Reply:
I like it for not needing to write the same values reference twice, but dislike that it is less obvious on how it behaves.

Reply:
Right. So if I see the `values.yaml` file, it looks like I'll actually get `loadBalancerIP: ""` in my Service object, which is confusing! Since k8s differentiates between unset values and `""` values, so should we - otherwise it can be pretty confusing. I think the mental model of `if a value is unspecified (null), it is not present in the target` is better than `if a value is falsey (by go template's definition of falsey) it will not be present in the target`. The latter also has exceptions, like storageClassName, making it even more confusing.

So ideally, I'd like to hold the mental model of 'if it is null in my values.yaml, it will not be present in the output Service object'.

Reply:
@yuvipanda I can make it so for strings/integers without problem I think, but if we apply the same principle to `{}` and `[]` I believe there will be trouble such as helm emitting warnings, broken assumptions by helm template logic, and potentially a schema that fail to capture otherwise easy to capture issues.

I suggest we apply the principle anyhow to our scalar strings and numbers, but don't try to apply it on falsy values for object `{}` and array `[]`, does that sound okay to you?

Reply:
That's perfect! I totally agree, @consideRatio :)

Reply:
Thanks for quickly iterating with me on this @yuvipanda!! I love it when I arrive at a concrete action point!

Reply:
It was a quite hard rule principle to follow without making exceptions. I was able to allow many strings to be null so our default values are now `null` rather than `""` by default, but plenty of booleans remain required to be "boolean" rather than "boolean or null" still.

__Example of complexity arisen:__
If helm templates have logic making an `eq` comparison between a variable and a string, then it will error if it holds a null value. And, since helm template logic will evaluate Y in statements like like `if X and Y`, one is forced to do `if X then if Y then ...` or `if X and (Y | default "")` etc.

Anyhow, things work okay now, don't want to change back, I don't see a perfect tradeoff with regards to these considerations.

---

Resolved by 2c1cc57